### PR TITLE
CRUD App review

### PR DESCRIPTION
**Pull request by mohit-sharma-87**
https://github.com/mongodb-developer/kotlin-driver-quick-start/blob/master/Readme.md

### CODE REVIEW

**Comment by johnlpage** [comment]
Does the Kotlin driver make I you use models like that, how do you support dynamic schema?
getenv is a very bad way to get your URI - it's not secure
I'm missing any error handling.
You don' check the connection works in setupConnection.
Why createCollection? and what on earth is maxDocuments? (Ah I see but the collection must be capped to use it)
should updateMultipleDocuments be updateAllDocuments - or add a filter?
Why are you using $regex in a simple tutorial - we don't really advise using $regex in most cases.
Paging with Skip and Limit is pretty controversial too although it's almost OK here.

File: src/main/kotlin/Setup.kt [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    }
+}
+
+fun setupConnection(
+    databaseName: String = "sample_restaurants",
+    connectionEnvVariable: String = "MONGODB_URI"
+):
+        MongoDatabase {
+    val connectString = if (System.getenv(connectionEnvVariable) != null) {
+        System.getenv(connectionEnvVariable)
```
Comment (mohit-sharma-87) [inline src/main/kotlin/Setup.kt:34]: Comment from @johnlpage : getenv is a very bad way to get your URI - it's not secure

File: src/main/kotlin/Setup.kt [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    } else {
+        "mongodb+srv://<username>:<enter your password>@cluster0.sq3aiau.mongodb" +
+                ".net/?retryWrites=true&w=majority"
+    }
+    val client = MongoClient.create(connectionString = connectString)
+    return client.getDatabase(databaseName = databaseName)
```
Comment (mohit-sharma-87) [inline src/main/kotlin/Setup.kt]: Comment form @johnlpage: You don' check the connection works in setupConnection.

File: src/main/kotlin/Setup.kt [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+suspend fun listAllCollection(database: MongoDatabase) {
+
+    print("Collection in this database are ---------------> ")
+    database.listCollectionNames().collect { print(" $it") }
+    println()
+}
+
+suspend fun createCollection(database: MongoDatabase) {
```
Comment (mohit-sharma-87) [inline src/main/kotlin/Setup.kt]: From @johnlpage - Why createCollection? and what on earth is maxDocuments? (Ah I see but the collection must be capped to use it)

File: src/main/kotlin/Update.kt [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    val queryParam = Filters.eq("restaurant_id", "restaurantId")
+    val updateParams = Updates.set("restaurant_id", Random.nextInt().toString())
+
+    collection.updateOne(filter = queryParam, update = updateParams).also {
+        println("Total docs modified ${it.matchedCount} and fields modified ${it.modifiedCount}")
+    }
+}
+
+
+suspend fun updateMultipleDocuments(db: MongoDatabase) {
```
Comment (mohit-sharma-87) [inline src/main/kotlin/Update.kt:27]: @johnlpage : should updateMultipleDocuments be updateAllDocuments - or add a filter?

File: src/main/kotlin/Read.kt [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    val collection = database.getCollection<Restaurant>(collectionName = "restaurants")
+    val queryParams = Filters
+        .and(
+            listOf(
+                eq(Restaurant::cuisine.name, "American"),
+                eq(Restaurant::borough.name, "Queens")
+            )
+        )
+
+    collection
```
Comment (mohit-sharma-87) [inline src/main/kotlin/Read.kt:58]: From @johnlpage : Paging with Skip and Limit is pretty controversial too although it's almost OK here.

File: src/main/kotlin/Delete.kt [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        deleteRestaurant(db)
+        deleteRestaurants(db)
+    }
+
+}
+
+suspend fun deleteRestaurant(db: MongoDatabase) {
+    val collection = db.getCollection<Restaurant>(collectionName = "restaurants")
+
+    val queryParams = Filters.regex(Restaurant::name.name, Pattern.compile("^Insert"))
```
Comment (mohit-sharma-87) [inline src/main/kotlin/Delete.kt]: From @johnlpage : Why are you using $regex in a simple tutorial - we don't really advise using $regex in most cases.

**Review by mohit-sharma-87 — COMMENTED** [review]
Comments from John

File: src/main/kotlin/Setup.kt (hunk 1 above)
Comment (dfreniche) [reply src/main/kotlin/Setup.kt:34]: @johnlpage I think security here is beyond the scope of this Quick Start example. Also, getting the connection string from an ENV variable is way better than hardcoding it. Which way of accessing these kind of secrets will you propose here? Thanks

File: src/main/kotlin/Setup.kt [hunk 7]
```diff
@@ -0,0 +1,58 @@
+import com.mongodb.client.model.CreateCollectionOptions
+import com.mongodb.kotlin.client.coroutine.MongoClient
+import com.mongodb.kotlin.client.coroutine.MongoDatabase
+import kotlinx.coroutines.runBlocking
+import java.util.*
+
+fun main() {
+    val databaseName = "sample_restaurants"
+    val database = setupConnection(databaseName = databaseName, "MONGODB_URI")
+    runBlocking {
```
Comment (dfreniche) [inline src/main/kotlin/Setup.kt:13]: This is blocking the main thread, right? Is there a way to do do this in a non-blocking way?

**Comment by dfreniche** [comment]
@mohit-sharma-87 this repo has a `master` branch that should be `main`. Better rename that.

**Comment by johnlpage** [comment]
Read from a file, or hard code.
Env is a dangerous mistake that keeps getting propogated.

**Comment by mohit-sharma-87** [comment]
> @mohit-sharma-87 this repo has a `master` branch that should be `main`. Better rename that.

Yep agreed and I planned to delete this branch post-merging.

File: src/main/kotlin/Read.kt [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    val sort = Sorts.ascending("restaurant_id")
+
+    collection
+        .find(filter = filters)
+        .sort(sort = sort)
+        .limit(200)
+        .collect {
+            println(it)
+        }
+
```
Comment (dfreniche) [inline src/main/kotlin/Read.kt]: Empty line

File: src/main/kotlin/Update.kt [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    val collection = db.getCollection<Restaurant>("restaurants")
+    val queryParam = Filters.empty()
+    val updateParams = Updates.set(Restaurant::cuisine.name, "Indian")
+
+    collection.updateMany(filter = queryParam, update = updateParams).also {
+        println("Total docs modified ${it.matchedCount} and fields modified ${it.modifiedCount}")
+    }
+
```
Comment (dfreniche) [inline src/main/kotlin/Update.kt]: Empty line

File: src/main/kotlin/Read.kt [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+            println(it)
+        }
+}
+
+suspend fun readWithIndex(database: MongoDatabase) {
+    val collection = database.getCollection<Restaurant>(collectionName = "restaurants")
+    val options = IndexOptions().apply {
+        this.name("restaurant_id_index")
+        this.background(true)
+
```
Comment (dfreniche) [inline src/main/kotlin/Read.kt]: Empty line

File: src/main/kotlin/Delete.kt (hunk 6 above)
Comment (mohit-sharma-87) [reply src/main/kotlin/Delete.kt]: @johnlpage: The object created earlier follows a particular naming pattern/

File: src/main/kotlin/Read.kt (hunk 5 above)
Comment (mohit-sharma-87) [reply src/main/kotlin/Read.kt:58]: Noted, added a note for the same in the blog.

File: src/main/kotlin/Setup.kt (hunk 1 above)
Comment (mohit-sharma-87) [reply src/main/kotlin/Setup.kt:34]: @johnlpage: I also have the same point of view as @dfreniche.

File: src/main/kotlin/Setup.kt (hunk 3 above)
Comment (mohit-sharma-87) [reply src/main/kotlin/Setup.kt]: @johnlpage: Just to demo how to create collection :) Are you suggesting something else or anything wrong with this approach?

File: src/main/kotlin/Update.kt (hunk 4 above)
Comment (mohit-sharma-87) [reply src/main/kotlin/Update.kt:27]: Done!

File: src/main/kotlin/Setup.kt (hunk 2 above)
Comment (mohit-sharma-87) [reply src/main/kotlin/Setup.kt]: @johnlpage: Could you add a suggestion?

File: src/main/kotlin/Delete.kt (hunk 6 above)
Comment (johnlpage) [reply src/main/kotlin/Delete.kt]: So? Still not good practise - what are you trying to show, why not use a range?

File: src/main/kotlin/Setup.kt (hunk 2 above)
Comment (johnlpage) [reply src/main/kotlin/Setup.kt]: Yup call client.admindb.runCommand('hello');

File: src/main/kotlin/Setup.kt (hunk 3 above)
Comment (johnlpage) [reply src/main/kotlin/Setup.kt]: I'm suggesting thart no-one should be calling createCollection when using MongoDB unless they have some unusual need (*like time series) so don't do it.

**Review by dfreniche — CHANGES_REQUESTED** [review]
Some comments

File: src/main/kotlin/Setup.kt (hunk 7 above)
Comment (mohit-sharma-87) [reply src/main/kotlin/Setup.kt:13]: I could but that would complicate the code. I am trying to keep it as simple as possible so people can quickly understand and start.

File: Readme.md [hunk 11]
```diff
@@ -0,0 +1,494 @@
...
+
+```kotlin
+collection.updateOne(filter = queryParam, update = updateParams).also {
+    println("Total docs modified ${it.matchedCount} and fields modified ${it.modifiedCount}")
+}
+```
+
+In the above example, we were already aware of which document we want to update, restaurant with an id `restauratantId` but in the real-world app
+that might not be the case. So then you would first look for the document and then update it (two operations for one task) which can lead to
+performance issues or race conditions.
```
Comment (nraboy) [inline Readme.md]: I don't think this is entirely true, or maybe it is phrased wrong.

If you're building an API and you are doing a PUT request to update a document, you're going to be passing some information that you know about the document, whether that be `_id` or something else. Sure maybe the user has done a GET prior in their frontend to find the document, but I think it would be rare to say you need two requests in a PUT endpoint. The `findOneAndUpdate` might also be a rare circumstance.

File: Readme.md [hunk 12]
```diff
@@ -0,0 +1,494 @@
...
+could be a Read operation, which we would do shortly or use [MongoDB compass](https://www.mongodb.com/products/compass) to view information.
+
+![MongoDB compass](https://images.contentstack.io/v3/assets/blt39790b633ee0d5a7/blt72fd049dd230ea55/648ae40e1fb2d38f0e495940/Screenshot_2023-06-15_at_11.12.23.png)
+
+
+---------------------
+
+## Read
+
+To read the information from the database, we can use the `find` operator and let's begin by reading a random document.
```
Comment (nraboy) [inline Readme.md]: A `find` operation with a `limit` does not return a random document. It returns the first document in the result set.

File: Readme.md [hunk 13]
```diff
@@ -0,0 +1,494 @@
...
+)
+```
+
+In the above code snippet, we used two annotations
+
+1. `@BsonId` which are used to represent to the as unique identity `_id` of a document.
+2. `@BsonProperty` which is used to create an alias for keys in the document like this case it represents `restaurant_id`.
+
+> Also, to note here our `Restaurant` data class here is exact replicate of restaurant document, but we can skip few fields like `grades` or
+`address` or make them optional while maintaining the ability to perform CRUD operation over it as MongoDB support flexible schema.
```
Comment (nraboy) [inline Readme.md:219]: Does an ignore happen by default in Kotlin? In C# you have to explicitly state which fields you want to ignore to prevent an error.

File: Readme.md [hunk 14]
```diff
@@ -0,0 +1,494 @@
...
+```kotlin
+suspend fun setupConnection(
+    databaseName: String = "sample_restaurants",
+    connectionEnvVariable: String = "MONGODB_URI"
+): MongoDatabase? {
+    val connectString = if (System.getenv(connectionEnvVariable) != null) {
+        System.getenv(connectionEnvVariable)
+    } else {
+        "mongodb+srv://mohitsharma:gq0Sj8aUXucHQtc2@cluster0.sq3aiau.mongodb.net/?retryWrites=true&w=majority"
+    }
```
Comment (nraboy) [inline Readme.md]: To avoid confusion and promote bad practice, I would remove this even knowing that this is an example.

File: Readme.md [hunk 15]
```diff
@@ -0,0 +1,494 @@
...
+## Adding MongoDB Kotlin driver
+
+Adding the driver to the project is simple and straightforward, just update the `dependencies` block in the build file i.e. `build.gradle` with
+these.
+
+```groovy
+dependencies {
+    implementation 'org.jetbrains.kotlinx:kotlinx-coroutines-core:1.6.4'
+    implementation 'org.mongodb:mongodb-driver-kotlin-coroutine:4.10.0-SNAPSHOT'
+}
```
Comment (nraboy) [inline Readme.md]: Might be worth explaining these dependencies so the reader isn't blindly adding things to their project.

File: Readme.md [hunk 16]
```diff
@@ -0,0 +1,494 @@
...
+
+[2]: https://www.mongodb.com/docs/atlas/sample-data/
+
+[3]: https://www.mongodb.com/cloud/atlas/register
+
+[4]: https://www.mongodb.com/docs/drivers/kotlin/coroutine/current/
+
+[5]: https://github.com/mongodb-developer/kotlin-driver-quick-start
+
+# Getting started with MongoDB for building backend Applications using the official Kotlin driver
```
Comment (nraboy) [inline Readme.md]: Since this is a series, it might be worth adding a "Part 1" to the title or being more explicit that this tutorial covers basic CRUD with Kotlin. This way users know what they are signing up for when they click the link.

File: Readme.md [hunk 17]
```diff
@@ -0,0 +1,494 @@
...
+
+> This is an introduction article on how to build backend application in Kotlin using official MongoDB Kotlin driver and [MongoDB Atlas][1].
```
Comment (nraboy) [inline Readme.md]: Rather than adding this line, can you give the article a proper introduction. Some ideas:

- Identify a problem that only Kotlin can solve and explain how this tutorial will help the reader solve that problem.
- Explain why you love Kotlin so much and highlight some of the things you're going to accomplish in the article.

Yes some readers will go straight to the code they can copy and paste, but others need a hook to know why they are going to continue reading. I personally fall into category 2.

File: Readme.md [hunk 18]
```diff
@@ -0,0 +1,494 @@
...
+And last, to aid our development activities, we would be using [Jetbrains IntelliJ IDEA (Community Edition)](https://www.jetbrains.
+com/idea/download/).
+
+---------------------
+
+## Kotlin driver vs Realm Kotlin SDK ?
+
+Before we start I would like to touch base on [Realm Kotlin SDK](https://www.mongodb.com/docs/realm/sdk/kotlin/), which is another framework by
+MongoDB in Kotlin that is used to create client applications using the MongoDB ecosystem and shouldn't be confused with this. [MongoDB Kotlin driver]
+[4] is used to create backends applications like RESTful applications which can be consumed by client apps.
```
Comment (nraboy) [inline Readme.md]: I'd rephrase this a bit. The drivers are appropriate to create backends, scripts, etc. Restricting the Kotlin driver to just RESTful applications would be a disservice to the driver.

File: Readme.md [hunk 19]
```diff
@@ -0,0 +1,494 @@
...
+val options = IndexOptions().apply {
+    this.name("restaurant_id_index")
+    this.background(true)
+}
+
+collection.createIndex(
+    keys = Indexes.ascending("restaurant_id"),
+    options = options
+)
+```
```
Comment (nraboy) [inline Readme.md:392]: Were indexes part of the request specification for this content? I'd leave them out if they were not part of the request, but it is up to you.

File: Readme.md [hunk 20]
```diff
@@ -0,0 +1,494 @@
...
+
+In general, there are two ways of updating any document(s) either
+
+* Use the **update** operation, which allows us to update specific fields of the matching documents without impacting the other fields.
+* Or **replace** operation which replaces the matching document with the new document.
+
+For this exercise, we would use the document we created earlier with create operation `{restaurant_id: "restaurantId"}` and update the
+`restaurant_id` with a more realistic value.
+
+First, we need to query the document using `Filters`, similar to Read operation earlier.
```
Comment (nraboy) [inline Readme.md]: Can we rephrase this?

The update operation isn't a two step approach. We're providing a filter and change criteria all in the same request. This currently reads like we are having users query for documents and then update documents.

File: Readme.md [hunk 21]
```diff
@@ -0,0 +1,494 @@
...
+## Delete
+
+Now let's see the last operation of our CRUD app i.e. Delete and start with how to delete a single document and here we would be
+using `findOneAndDelete` instead `deleteOne` which also returns the deleted document as output.
+
+```kotlin
+val collection = db.getCollection<Restaurant>(collectionName = "restaurants")
+val queryParams = Filters.eq("restaurant_id", "restaurantId")
+
+collection.findOneAndDelete(filter = queryParams).also {
```
Comment (nraboy) [inline Readme.md:466]: The `findOneAndDelete` and `findOneAndUpdate` are cool, but I don't know if I'd include them in a getting started guide. Probably don't want to overwhelm the user with commands.

File: Readme.md [hunk 22]
```diff
@@ -0,0 +1,494 @@
...
+}
+```
+
+---------------------
+
+## Summary
+
+With this blog post, we have covered all the basic operations and have also seen how easily we can leverage the simplicity of Kotlin with the MongoDB
+Kotlin driver. Thank you for reading and hopefully you find this article informative! The complete source code of the app can be found on
+[GitHub][5].
```
Comment (nraboy) [inline Readme.md:499]: Do you want to hint at the next tutorials in the series to come?

Comment (mohit-sharma-87) [reply Readme.md:499]: As discussed earlier, this is not a series.

File: Readme.md (hunk 21 above)
Comment (mohit-sharma-87) [reply Readme.md:466]: I would prefer to keep it.

File: Readme.md (hunk 16 above)
Comment (mohit-sharma-87) [reply Readme.md]: As discussed earlier, this is a standalone article.

File: Readme.md (hunk 14 above)
Comment (mohit-sharma-87) [reply Readme.md]: First thank you for flagging this, this got leaked accidentally due to the last min of code refactoring and copy-pasting. 

I have mentioned just below it and highlighted that it's not a good practice to do this and at the same time don't want to add a barrier for first-time users to run the code. Will also add a code comment to twice highlight not a good practice.

File: Readme.md (hunk 12 above)
Comment (mohit-sharma-87) [reply Readme.md]: Yes, I am aware of that. But I am talking from the user's perspective, not from the code. Happy to rephrase it if you feel it's giving the wrong message.

Comment (mohit-sharma-87) [reply Readme.md]: rephrased.

Also, just below code I already had an explain what we are doing to further remove confusion/misinformation about combination of find & limit. 

```
`find` operator returns a list of results but since we are only interested in a single document, we can use `limit` operator in conjunction to
limit our result set, in this case, it would be a single document.
```

File: Readme.md (hunk 15 above)
Comment (mohit-sharma-87) [reply Readme.md]: It's a standard practice for Kotlin/Java developer, also added a code comment to make it super clear.

File: Readme.md (hunk 19 above)
Comment (mohit-sharma-87) [reply Readme.md:392]: Indirectly yes.

File: Readme.md (hunk 13 above)
Comment (mohit-sharma-87) [reply Readme.md:219]: Sorry, I think so you understood it around way round.  We don't need to have explicate of document into the code to perform crud operation. 

`@Ignore ` is think so what you are talking about which allows you to ignore fields in syncing.
